metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/data_resolver.py

```diff
@@ -299,6 +308,267 @@ def get_report(self, workspace: Workspace, report_id: str) -> Optional[Report]:
 
         return reports[0]
 
+    def parse_sub_entity_level_usage_metrics_result(
+        self, results: List[Dict], user_stats_key_as_guid: bool
+    ) -> Dict[str, Dict[str, Dict[str, UsageStat]]]:
+        """
+        Return sub entity level usage metrics as Dict[<entity_id>, Dict[<sub_entity_id>, Dict[<date>, UsageStat]]].
+        """
+        usage_stats: Dict[str, Dict[str, Dict[str, UsageStat]]] = defaultdict()
+        for row in results:
+            entity_id = row[Constant.ENTITY_ID].lower()
+            sub_entity_id = row[Constant.SUB_ENTITY_ID].lower()
+            date = row[Constant.DATE]
+            user_id = row[Constant.USER_ID].lower()
+            views_count = row[Constant.VIEWS_COUNT]
+
+            if entity_id not in usage_stats:
+                usage_stats[entity_id] = {
+                    sub_entity_id: {
+                        date: UsageStat(
+                            user_stats_key_as_guid,
+                            {user_id: UserUsageStat(views_count)},
+                        )
+                    }
+                }
+            elif sub_entity_id not in usage_stats[entity_id]:
+                usage_stats[entity_id][sub_entity_id] = {
+                    date: UsageStat(
+                        user_stats_key_as_guid, {user_id: UserUsageStat(views_count)}
+                    )
+                }
+            elif date not in usage_stats[entity_id][sub_entity_id]:
+                usage_stats[entity_id][sub_entity_id][date] = UsageStat(
+                    user_stats_key_as_guid, {user_id: UserUsageStat(views_count)}
+                )
+            elif (
+                user_id
+                not in usage_stats[entity_id][sub_entity_id][date].userUsageStats
+            ):
+                usage_stats[entity_id][sub_entity_id][date].userUsageStats[
+                    user_id
+                ] = UserUsageStat(views_count)
+            else:
+                usage_stats[entity_id][sub_entity_id][date].userUsageStats[
+                    user_id
+                ].viewsCount = (
+                    usage_stats[entity_id][sub_entity_id][date]
+                    .userUsageStats[user_id]
+                    .viewsCount
+                    + views_count
+                )
```

Comment:
What does this represent ? Is it possible to have multiple rows for same  (entity, subentity, user, date)  combination ? Can we not push this down in Dax query so that it returns single value per (entity, subentity, user, date) ?

Reply:
Done